File: Problems/208/main.go
package main

import "fmt"

type TrieNode struct {
	decendents  [26]*TrieNode
	isEndofNode bool
}

type Trie struct {
	head *TrieNode
}

func main() {
	obj := Constructor()
	obj.Insert("apple")
	fmt.Println(obj.Search("apple"))
	fmt.Println(obj.Search("app"))
	fmt.Println(obj.StartsWith("app"))
	obj.Insert("app")
	fmt.Println(obj.Search("app"))
}

// How to Deal with Ending Node

func Constructor() Trie {
	return Trie{
		head: &TrieNode{},
	}
}

func (this *Trie) Insert(word string) {
	cur := this.head
	for _, r := range word {
		if cur.decendents[r-'a'] == nil {
			cur.decendents[r-'a'] = &TrieNode{}
		}
		cur = cur.decendents[r-'a']
	}
	cur.isEndofNode = true
}

func (this *Trie) Search(word string) bool {
	cur := this.head
	for _, r := range word {
		if cur.decendents[r-'a'] == nil {
			return false
		}
		cur = cur.decendents[r-'a']
	}
	return cur.isEndofNode
}

func (this *Trie) StartsWith(prefix string) bool {
	cur := this.head
	for _, r := range prefix {
		if cur.decendents[r-'a'] == nil {
			return false
		}
		cur = cur.decendents[r-'a']
	}
	return true
}

File: Problems/208/main.cc
#include <vector>
#include <iostream>
#include <string>
#include <map>
#include <stack>
#include <algorithm>
#include <numeric>
#include <unordered_set>
#include <unordered_map>
#include <queue>
#include <functional>
#include <sstream>
#include <set>
#include <cctype>

using namespace std;

struct Node
{
    vector<Node *> childs;
    bool isEnd;

    Node()
    {
        isEnd = false;
        childs.resize(26, nullptr);
    }
};

class Trie
{
public:
    Node *root;
    Trie()
    {
        root = new Node();
    }

    void insert(string word)
    {
        Node *cur = root;
        for (auto &w : word)
        {
            if (!cur->childs[w - 'a'])
                cur->childs[w - 'a'] = new Node();
            cur = cur->childs[w - 'a'];
        }
        cur->isEnd = true;
    }

    bool search(string word)
    {
        Node *cur = root;
        for (auto &w : word)
        {
            if (!cur->childs[w - 'a'])
                return false;
            cur = cur->childs[w - 'a'];
        }
        return cur->isEnd;
    }

    bool startsWith(string prefix)
    {
        Node *cur = root;
        for (auto &w : prefix)
        {
            if (!cur->childs[w - 'a'])
                return false;
            cur = cur->childs[w - 'a'];
        }
        return true;
    }
};

/**
 * Your Trie object will be instantiated and called as such:
 * Trie* obj = new Trie();
 * obj->insert(word);
 * bool param_2 = obj->search(word);
 * bool param_3 = obj->startsWith(prefix);
 */
